### ISSUE TITLE
Memory leak in etcd-manager

### ISSUE THREAD

**Issue by joshbohde** [kubernetes/kops#6875]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

v1.11.1

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.0", GitCommit:"641856db18352033a0d96dbc99153fa3b27298e5", GitTreeState:"clean", BuildDate:"2019-03-25T15:53:57Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.8", GitCommit:"4e209c9383fa00631d124c8adcc011d617339b3c", GitTreeState:"clean", BuildDate:"2019-02-28T18:40:05Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
```

**3. What cloud provider are you using?**

AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

Updated our test cluster to use etcd-manager, using the following manifest snippet:

```
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master-us-west-2a
      name: a
    name: main
    provider: Manager
    version: 2.2.1
  - etcdMembers:
    - instanceGroup: master-us-west-2a
      name: a
    name: events
    provider: Manager
    version: 2.2.1
```

**5. What happened after the commands executed?**

`etcd-manager` leaks memory at the rate of ~41MB per second on a mostly idle cluster, until the process OOMs. Once this happens, the `kube-apiserver` is unavailable until `etcd-manager` has restarted.

**6. What did you expect to happen?**

Using the previous etcd method does not leak memory, with an RSS of ~30MB on an idle cluster. 

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

**9. Anything else do we need to know?**

**Comment by justinsb** [comment]
Sorry about the issue @joshbohde, and thank you for reporting.

I'm not able to reproduce - for me memory as reported in e.g. `ps aux` is steady, at about 30MB.  40MB per second is _huge_, so I think something very abnormal must be happening.

Can you confirm the image of etcd-manager that you're running - based on the description I'm guessing it is this one:

```
>kubectl describe pod -n kube-system etcd-manager-main | grep Image
    Image:         kopeio/etcd-manager:3.0.20190224
    Image ID:      docker-pullable://kopeio/etcd-manager@sha256:4e4822c5f5ebe4b692dca46c11876a8dac967916e03f9e96db99939c78e76f8d
```

Also, are you doing something more sophisticated than looking at `ps aux`?  I also checked `pmap`, but perhaps I am using the wrong tooling?

Are there any clues in `/var/log/etcd.log` on your master?  You might want to exclude the ping messages for your own sanity e.g. ` tail -f /var/log/etcd.log | grep -v got.ping`.  For me, I observe this in a loop:

```
I0509 05:29:06.382911    3489 controller.go:173] starting controller iteration
I0509 05:29:06.382968    3489 controller.go:183] peers: [peer{id:"etcd-a" endpoints:"172.20.47.21:3996" }]
I0509 05:29:06.383028    3489 controller.go:269] I am leader with token "lnH__rw3KFXD91RWjDbWhg"
I0509 05:29:06.385199    3489 controller.go:276] etcd cluster state: etcdClusterState
  members:
    {"name":"etcd-a","peerURLs":["http://etcd-a.internal.etcdmanager.awsdata.com:2380"],"endpoints":["http://etcd-a.internal.etcdmanager.awsdata.com:4001"],"ID":"46052d31767c767d"}
  peers:
    etcdClusterPeerInfo{peer=peer{id:"etcd-a" endpoints:"172.20.47.21:3996" }, info=cluster_name:"etcd" node_configuration:<name:"etcd-a" peer_urls:"http://etcd-a.internal.etcdmanager.awsdata.com:2380" client_urls:"http://etcd-a.internal.etcdmanager.awsdata.com:4001" quarantined_client_urls:"http://etcd-a.internal.etcdmanager.awsdata.com:3994" > etcd_state:<cluster:<cluster_token:"rZG7aSYKa7WOMBMfx2dAFg" nodes:<name:"etcd-a" peer_urls:"http://etcd-a.internal.etcdmanager.awsdata.com:2380" client_urls:"http://etcd-a.internal.etcdmanager.awsdata.com:4001" quarantined_client_urls:"http://etcd-a.internal.etcdmanager.awsdata.com:3994" > > etcd_version:"2.2.1" > }
I0509 05:29:06.385269    3489 controller.go:277] etcd cluster members: map[46052d31767c767d:{"name":"etcd-a","peerURLs":["http://etcd-a.internal.etcdmanager.awsdata.com:2380"],"endpoints":["http://etcd-a.internal.etcdmanager.awsdata.com:4001"],"ID":"46052d31767c767d"}]
I0509 05:29:06.385373    3489 controller.go:615] sending member map to all peers: members:<name:"etcd-a" dns:"etcd-a.internal.etcdmanager.awsdata.com" addresses:"172.20.47.21" > 
I0509 05:29:06.385590    3489 etcdserver.go:222] updating hosts: map[172.20.47.21:[etcd-a.internal.etcdmanager.awsdata.com]]
I0509 05:29:06.385607    3489 hosts.go:81] hosts update: primarymap[172.20.47.21:[etcd-a.internal.etcdmanager.awsdata.com]], fallbacks=map[etcd-a.internal.etcdmanager.awsdata.com:[172.20.47.21 172.20.47.21]], final=map[172.20.47.21:[etcd-a.internal.etcdmanager.awsdata.com]]
I0509 05:29:06.385872    3489 commands.go:22] not refreshing commands - TTL not hit
I0509 05:29:06.385890    3489 s3fs.go:219] Reading file "s3://clusters.awsdata.com/etcdmanager.awsdata.com/backups/etcd/main/control/etcd-cluster-created"
I0509 05:29:06.412355    3489 controller.go:369] spec member_count:1 etcd_version:"2.2.1" 
I0509 05:29:06.412895    3489 reconcile_configuration.go:73] TLS configuration is disabled, won't enable TLS
I0509 05:29:06.413104    3489 controller.go:531] controller loop complete
```

**Comment by joshbohde** [comment]
I misspoke, I'm seeing a leak of ~41MB per _hour_. Sorry for the confusion!

We're using the following image:

```
    Image:         kopeio/etcd-manager:3.0.20190224
    Image ID:      docker-pullable://kopeio/etcd-manager@sha256:4e4822c5f5ebe4b692dca46c11876a8dac967916e03f9e96db99939c78e76f8d
```

We've also tried using `kopeio/etcd-manager:3.0.20190328`, which exhibits the same behavior.  

We've since tested this to a 3 master cluster. It exhibits the same memory leak, but at different rates between the leader and the followers. Here is the configuration:

```
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master-us-west-2a
      name: us-west-2a
    - instanceGroup: master-us-west-2b
      name: us-west-2b
    - instanceGroup: master-us-west-2c
      name: us-west-2c
    name: main
    provider: Manager
    version: 2.2.1
  - etcdMembers:
    - instanceGroup: master-us-west-2a
      name: us-west-2a
    - instanceGroup: master-us-west-2b
      name: us-west-2b
    - instanceGroup: master-us-west-2c
      name: us-west-2c
    name: events
    provider: Manager
    version: 2.2.1
```

The way we found this leak was through our metrics collected via cadvisor, using the following Prometheus query:

```
avg (container_memory_rss{image!="",name=~"^k8s_.*", pod_name=~"^etcd-manager-main-.*", namespace="kube-system"}) by (pod_name, container_name)
```
Here's a graph of this immediately following the deployment: ![container_memory_rss](https://d2ddoduugvun08.cloudfront.net/items/3p1n1R3k2l1C0c0t2H1F/%5B2c128245ab0233ba75ac61959b62af6a%5D_Image%25202019-05-09%2520at%252010.44.36%2520AM.png?X-CloudApp-Visitor-Id=3203319)


Our controller loop log looks like the following: 

```
I0503 20:39:50.216954    3476 controller.go:173] starting controller iteration
I0503 20:39:50.216989    3476 controller.go:183] peers: [peer{id:"etcd-us-west-2a" endpoints:"172.0.0.0:3996" } peer{id:"etcd-us-west-2b" endpoints:"172.0.0.1:3996" } peer{id:"etcd-us-west-2c" endpoints:"172.0.0.3:3996" }]
I0503 20:39:50.217033    3476 controller.go:269] I am leader with token "riroh84_YfxS2_G_vAH8SA"
I0503 20:39:50.226022    3476 controller.go:276] etcd cluster state: etcdClusterState
  members:
    {"name":"etcd-us-west-2c","peerURLs":["http://etcd-us-west-2c.internal.k8.example.com:2380"],"endpoints":["http://etcd-us-west-2c.internal.k8.example.com:4001"],"ID":"f9a49842990d5787"}
    {"name":"etcd-us-west-2a","peerURLs":["http://etcd-us-west-2a.internal.k8.example.com:2380"],"endpoints":["http://etcd-us-west-2a.internal.k8.example.com:4001"],"ID":"9b394fe4e88d4815"}
    {"name":"etcd-us-west-2b","peerURLs":["http://etcd-us-west-2b.internal.k8.example.com:2380"],"endpoints":["http://etcd-us-west-2b.internal.k8.example.com:4001"],"ID":"cde2c58a511217ad"}
  peers:
    etcdClusterPeerInfo{peer=peer{id:"etcd-us-west-2a" endpoints:"172.0.0.0:3996" }, info=cluster_name:"etcd" node_configuration:<name:"etcd-us-west-2a" peer_urls:"http://etcd-us-west-2a.internal.k8.example.com:2380" client_urls:"http://etcd-us-west-2a.internal.k8.example.com:4001" quarantined_client_urls:"http://etcd-us-west-2a.internal.k8.example.com:3994" > etcd_state:<cluster:<desired_cluster_size:3 cluster_token:"etcd-cluster-token-etcd" nodes:<name:"etcd-us-west-2a" peer_urls:"http://etcd-us-west-2a.internal.k8.example.com:2380" client_urls:"http://0.0.0.0:4001" quarantined_client_urls:"http://0.0.0.0:3994" > nodes:<name:"etcd-us-west-2b" peer_urls:"http://etcd-us-west-2b.internal.k8.example.com:2380" client_urls:"http://0.0.0.0:4001" quarantined_client_urls:"http://0.0.0.0:3994" > nodes:<name:"etcd-us-west-2c" peer_urls:"http://etcd-us-west-2c.internal.k8.example.com:2380" client_urls:"http://0.0.0.0:4001" quarantined_client_urls:"http://0.0.0.0:3994" > > etcd_version:"2.2.1" > }
    etcdClusterPeerInfo{peer=peer{id:"etcd-us-west-2b" endpoints:"172.0.0.1:3996" }, info=cluster_name:"etcd" node_configuration:<name:"etcd-us-west-2b" peer_urls:"http://etcd-us-west-2b.internal.k8.example.com:2380" client_urls:"http://etcd-us-west-2b.internal.k8.example.com:4001" quarantined_client_urls:"http://etcd-us-west-2b.internal.k8.example.com:3994" > etcd_state:<cluster:<desired_cluster_size:3 cluster_token:"etcd-cluster-token-etcd" nodes:<name:"etcd-us-west-2a" peer_urls:"http://etcd-us-west-2a.internal.k8.example.com:2380" client_urls:"http://0.0.0.0:4001" quarantined_client_urls:"http://0.0.0.0:3994" > nodes:<name:"etcd-us-west-2b" peer_urls:"http://etcd-us-west-2b.internal.k8.example.com:2380" client_urls:"http://0.0.0.0:4001" quarantined_client_urls:"http://0.0.0.0:3994" > nodes:<name:"etcd-us-west-2c" peer_urls:"http://etcd-us-west-2c.internal.k8.example.com:2380" client_urls:"http://0.0.0.0:4001" quarantined_client_urls:"http://0.0.0.0:3994" > > etcd_version:"2.2.1" > }
    etcdClusterPeerInfo{peer=peer{id:"etcd-us-west-2c" endpoints:"172.0.0.3:3996" }, info=cluster_name:"etcd" node_configuration:<name:"etcd-us-west-2c" peer_urls:"http://etcd-us-west-2c.internal.k8.example.com:2380" client_urls:"http://etcd-us-west-2c.internal.k8.example.com:4001" quarantined_client_urls:"http://etcd-us-west-2c.internal.k8.example.com:3994" > etcd_state:<cluster:<desired_cluster_size:3 cluster_token:"etcd-cluster-token-etcd" nodes:<name:"etcd-us-west-2a" peer_urls:"http://etcd-us-west-2a.internal.k8.example.com:2380" client_urls:"http://0.0.0.0:4001" quarantined_client_urls:"http://0.0.0.0:3994" > nodes:<name:"etcd-us-west-2b" peer_urls:"http://etcd-us-west-2b.internal.k8.example.com:2380" client_urls:"http://0.0.0.0:4001" quarantined_client_urls:"http://0.0.0.0:3994" > nodes:<name:"etcd-us-west-2c" peer_urls:"http://etcd-us-west-2c.internal.k8.example.com:2380" client_urls:"http://0.0.0.0:4001" quarantined_client_urls:"http://0.0.0.0:3994" > > etcd_version:"2.2.1" > }
I0503 20:39:50.226191    3476 controller.go:277] etcd cluster members: map[9b394fe4e88d4815:{"name":"etcd-us-west-2a","peerURLs":["http://etcd-us-west-2a.internal.k8.example.com:2380"],"endpoints":["http://etcd-us-west-2a.internal.k8.example.com:4001"],"ID":"9b394fe4e88d4815"} cde2c58a511217ad:{"name":"etcd-us-west-2b","peerURLs":["http://etcd-us-west-2b.internal.k8.example.com:2380"],"endpoints":["http://etcd-us-west-2b.internal.k8.example.com:4001"],"ID":"cde2c58a511217ad"} f9a49842990d5787:{"name":"etcd-us-west-2c","peerURLs":["http://etcd-us-west-2c.internal.k8.example.com:2380"],"endpoints":["http://etcd-us-west-2c.internal.k8.example.com:4001"],"ID":"f9a49842990d5787"}]
I0503 20:39:50.226215    3476 controller.go:615] sending member map to all peers: members:<name:"etcd-us-west-2a" dns:"etcd-us-west-2a.internal.k8.example.com" addresses:"172.0.0.0" > members:<name:"etcd-us-west-2b" dns:"etcd-us-west-2b.internal.k8.example.com" addresses:"172.0.0.1" > members:<name:"etcd-us-west-2c" dns:"etcd-us-west-2c.internal.k8.example.com" addresses:"172.0.0.3" > 
I0503 20:39:50.226445    3476 etcdserver.go:222] updating hosts: map[172.0.0.0:[etcd-us-west-2a.internal.k8.example.com] 172.0.0.1:[etcd-us-west-2b.internal.k8.example.com] 172.0.0.3:[etcd-us-west-2c.internal.k8.example.com]]
I0503 20:39:50.226484    3476 hosts.go:81] hosts update: primarymap[172.0.0.0:[etcd-us-west-2a.internal.k8.example.com] 172.0.0.1:[etcd-us-west-2b.internal.k8.example.com] 172.0.0.3:[etcd-us-west-2c.internal.k8.example.com]], fallbacks=map[etcd-us-west-2b.internal.k8.example.com:[172.0.0.1 172.0.0.1] etcd-us-west-2a.internal.k8.example.com:[172.0.0.0 172.0.0.0] etcd-us-west-2c.internal.k8.example.com:[172.0.0.3 172.0.0.3]], final=map[172.0.0.3:[etcd-us-west-2c.internal.k8.example.com] 172.0.0.0:[etcd-us-west-2a.internal.k8.example.com] 172.0.0.1:[etcd-us-west-2b.internal.k8.example.com]]
I0503 20:39:50.229134    3476 commands.go:22] not refreshing commands - TTL not hit
I0503 20:39:50.229163    3476 s3fs.go:219] Reading file "s3://example-kops-state-store/k8.example.com/backups/etcd/main/control/etcd-cluster-created"
I0503 20:39:50.254241    3476 controller.go:369] spec member_count:3 etcd_version:"2.2.1" 
I0503 20:39:50.254295    3476 reconcile_configuration.go:73] TLS configuration is disabled, won't enable TLS
I0503 20:39:50.254306    3476 controller.go:531] controller loop complete
```

Here is the htop of the etd-manager-main container process tree on a follower after it's been running for a few days:

![htop](https://d2ddoduugvun08.cloudfront.net/items/1f3Z3L0O1f0i180Y3T13/%5Bd9a9e6707d412dd88d0d728523830303%5D_Image+2019-05-09+at+11.28.19+AM.png?X-CloudApp-Visitor-Id=3203319&v=1aab2b92)

The resident size of this follower etcd is 910m, but nothing else stands out to me. We're continuing to investigate, but the leader has become unresponsive. 

We've couldn't find a documented way of overriding the flags etcd-manager passes to etcd. If we had this, we could enable pprof endpoints in order to debug the etcd's heap.

**Comment by marek-obuchowicz** [comment]
It happens for me as well. I ended up with same versions (etcd 2.2.1, etcd-manager 3.0.20190224) on AWS cluster and we have similiar behaviour - memleak of etcd - which happens always on one machine (master node).

When I was checking etcd2 isues on github, there was actually a lot of memleaks reported. The recommendation to mitigate the issue is to switch to etcd3.

When I was using kops-1.12 alpha, migration to etcd3 was a complete disaster and it caused loss of all data in etcd and k8s down. I will give another try to kops 1.12 within next days to see if it finally fixes the memleak issue.

**Comment by joshbohde** [comment]
Investigating this more, this seems to be caused by a file descriptor leak. 

All etcd instances running under etcd-manager now leak file descriptors, at roughly the same rate they are leaking memory. Leaders leak memory and file descriptors more. ![graph](https://cl.ly/3b92777a5313/Image%2525202019-05-20%252520at%25252011.40.50%252520AM.png)

Looking into lsof output, there are a ton of TCP connections in an established state. `etcd-us-west-2a` is the leader. ![lsof output](https://cl.ly/f9556611cc82/%255B821648c2b236d4da853a119969075403%255D_Image%2525202019-05-20%252520at%25252011.40.34%252520AM.png)

**Comment by Pajk** [comment]
Seeing the same, etcd manager memory is not really steady.

Kops 1.12, k8s 1.12.8, etcd 3.2.24

![image](https://user-images.githubusercontent.com/624416/59202173-357c2c00-8b9c-11e9-8d29-d02881d57aff.png)

**Comment by lawrencedudley** [comment]
We're seeing this - the normal bumps seem normal but now the spikes are way bigger.

We were peaking around 900MB, now they're up to 4GB and eventually it causes an OOM.

**Comment by VihanAgarwal97** [comment]
I'm on @joshbohde's team, and our peak reduced from 2.1 Gb to ~ 300 Mb after we upgraded to k8s 1.12.8, etcd 3.2.24 and etcd-manager 3.0.20190513. However, we still see the same pattern of memory usage.

**Comment by lawrencedudley** [comment]
Here's some stats. We haven't changed anything to make this happen:

avg (container_memory_rss{image!="",name=~"^k8s_.*", pod_name=~"^etcd-manager-main-.*", namespace="kube-system"}) by (pod_name, container_name)

<img width="1349" alt="Screen Shot 2019-06-19 at 07 47 06" src="https://user-images.githubusercontent.com/272317/59742889-970c5c80-9266-11e9-8706-b7b1a0dd8e91.png">

This feels like it can only be related to history compaction in etcd but we haven't made any changes in our cluster that would cause this many more keys to be written.

kopeio/etcd-manager:3.0.20190516 with Kubernetes 1.12.9

**Comment by lawrencedudley** [comment]
`spec:
  api:
    loadBalancer:
      type: Public
  authorization:
    rbac: {}
  channel: stable
  cloudProvider: aws
  configBase: xxx
  dnsZone: xxx
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master-eu-west-1a
      name: a
    - instanceGroup: master-eu-west-1b
      name: b
    - instanceGroup: master-eu-west-1c
      name: c
    name: main
    provider: Manager
    version: 3.3.10
  - etcdMembers:
    - instanceGroup: master-eu-west-1a
      name: a
    - instanceGroup: master-eu-west-1b
      name: b
    - instanceGroup: master-eu-west-1c
      name: c
    name: events
    provider: Manager
    version: 3.3.10`

The version also gets completely ignored - etcd manager seems to start /opt/etcd-v3.2.24-linux-amd64/etcd regardless.

**Comment by lawrencedudley** [comment]
Correction: It is now running 3.3.10. Will monitor to see if the memory usage flattens out at all.

Also we've seen an increase in read/write latency and associated increase in read/write (mostly write) bandwidth.

Is there a chance that this issue isn't actually etcd related? As in, something in the cluster has increased the amount of work etcd has to do.

**Comment by nirousseau** [comment]
Hello Guys,

I'm also issuing the same strange memory usage...

Etcd consumes up to 14GB

![__KUBE](https://user-images.githubusercontent.com/48907457/61442398-d3a3a480-a947-11e9-9383-10f368cf356a.PNG)

etcd 3.2.24

etcd is deployed using RKE 0.2.4, Rancher 2.2.4

I will try to provide more logs asap

**Comment by lawrencedudley** [comment]
Some more digging - I think this is related to the memory leak in grpc
proxy which is fixed in etcd 3.3.11.

Annoyingly etcd-manager doesn't support >3.3.10 currently so I can't verify
but I'm fairly sure it's the same thing.

Kind Regards,

Lawrence Dudley
Director

**Comment by juris** [comment]
Same here. Kops 1.12.3, Kubernetes 1.11.10, etcd 2.2.1, etcd-manager 3.0.20190801
